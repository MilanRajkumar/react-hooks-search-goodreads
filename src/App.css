html * {
  font-size: 16px;
  font-family: sans-serif;
}

.App {
  text-align: center;
  position: relative;
}

#root {
  height: calc(100vh);
  background-image: url('./assets/bg.jpeg');
}

.card {
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  background: #fff;
}

fieldset {
  border: none;
  height: 2rem;
}
.f2 {
  font-size: 2rem;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
