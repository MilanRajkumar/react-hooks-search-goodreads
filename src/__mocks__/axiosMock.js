export default {
  useBookNameFetch: jest.fn().mockResolvedValue({
    books: [],
  }),
};
