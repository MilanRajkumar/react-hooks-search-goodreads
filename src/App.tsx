import React, { useState, useEffect, useRef } from 'react';

import './components/books/books.css';
import './App.css';
import SearchBooks from './components/books/searchBooks';
import { useBookNameFetch } from './services';
import { key } from './constants';
import { RouteComponentProps } from 'react-router';

const App: React.FC<RouteComponentProps> = ({ history }) => {
  let textInput = useRef<HTMLInputElement>(null);

  const [url, setUrl] = useState(`/search.xml?key=${key}&q=${'Game'}`);
  const [books, setBooks] = useState<any>([]);

  useEffect(() => {
    const node = textInput.current;
    node ? node.focus() : '';
    let fetchBooks = async () => {
      const result = await useBookNameFetch(url);
      setBooks(result);
    };

    fetchBooks();
  }, [url]);

  const onSearchBooks = (text: string) => {
    const url = `/search.xml?key=${key}&q=${text}`;
    setUrl(url);
  };

  const listPlaceHolder = () => {
    return (
      <div style={{ width: '100%', textAlign: 'center', marginTop: '4rem', fontSize: '2rem' }}>
        <span style={{ color: '#fff' }} data-testid="imageListBeforeId">
          Your books will be shown here
        </span>
      </div>
    );
  };

  const naviagateToBookDetails = (id: number) => {
    history.push({
      pathname: '/book-details',
      state: {
        bookId: `${id}`,
      },
    });
  };

  return (
    <div className="App">
      <SearchBooks onSubmit={onSearchBooks} ref={textInput} />
      {books.length ? (
        <div className="gallery-wrapper">
          <div className="gallery" data-testId="imageListAfterId">
            {books.length
              ? books.map((item: any, i: number) => (
                  <img
                    key={item.id._text + i.toString()}
                    src={item.best_book.image_url._text}
                    className="gallery__img"
                    onClick={() => naviagateToBookDetails(item.best_book.id._text)}
                  />
                ))
              : ''}
          </div>
        </div>
      ) : (
        listPlaceHolder()
      )}
    </div>
  );
};

export default App;
