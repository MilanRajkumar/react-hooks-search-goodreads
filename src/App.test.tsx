import React from 'react';
import { render, cleanup, waitForElement, fireEvent, act } from 'react-testing-library';
import 'jest-dom/extend-expect';
import SearchBox from './components/books/searchBooks';
import App from './App';
import axiosMock from './__mocks__/axiosMock';

afterEach(cleanup);
it('render search box', () => {
  const { asFragment } = render(<SearchBox onSubmit={() => {}} />);
  expect(asFragment).toMatchSnapshot();
});

it('render image list placeholder', () => {
  const { getByTestId } = render(<App />);
  expect(getByTestId('imageListBeforeId')).toHaveTextContent('Your books will be shown here');
});

it('render image list', async () => {
  axiosMock.useBookNameFetch.mockResolvedValueOnce({
    books: [
      {
        id: {
          _type: 'integer',
          __text: '10523104',
        },
        books_count: {
          _type: 'integer',
          __text: '4',
        },
        ratings_count: {
          _type: 'integer',
          __text: '741',
        },
        text_reviews_count: {
          _type: 'integer',
          __text: '30',
        },
        original_publication_year: {
          _type: 'integer',
          __text: '2009',
        },
        original_publication_month: {
          _type: 'integer',
          __text: '12',
        },
        original_publication_day: {
          _type: 'integer',
          __text: '9',
        },
        average_rating: '3.95',
        best_book: {
          id: {
            _type: 'integer',
            __text: '7734468',
          },
          title: "Ender's Game: War of Gifts",
          author: {
            id: {
              _type: 'integer',
              __text: '236462',
            },
            name: 'Jake Black',
          },
          image_url:
            'https://s.gr-assets.com/assets/nophoto/book/111x148-bcc042a9c91a29c1d680899eff700a03.png',
          small_image_url:
            'https://s.gr-assets.com/assets/nophoto/book/50x75-a91bf249278a81aabab721ef782c4a74.png',
          _type: 'Book',
        },
      },
    ],
  });
  const { getByTestId } = render(<App />);
  fireEvent.click(getByTestId('searchButtonId'));
  const imageList = await waitForElement(() => {
    getByTestId('imageListAfterId');
  });
  expect(axiosMock.useBookNameFetch).toHaveBeenCalledTimes(9);
  // expect(axiosMock.useBookNameFetch).toHaveBeenCalledWith(url);
});
