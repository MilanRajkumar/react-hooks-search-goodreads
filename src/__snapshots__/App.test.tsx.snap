// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render search box 1`] = `[Function]`;
